### PR TITLE
nvcc wrapper: Add -ccbin only if not already supplied

### PR DESCRIPTION
**Pull request by mbargull**
Checklist
* [x] Used a [personal fork of the feedstock to propose changes](https://conda-forge.org/docs/maintainer/updating_pkgs.html#forking-and-pull-requests)
* [x] Bumped the build number (if the version is unchanged)
* [x] [Re-rendered]( https://conda-forge.org/docs/maintainer/updating_pkgs.html#rerendering-feedstocks ) with the latest `conda-smithy` (Use the phrase <code>@<space/>conda-forge-admin, please rerender</code> in a comment in this PR for automated rerendering)
* [ ] Ensured the license file is being packaged.

Closes gh-34
Probably fixes https://github.com/conda-forge/faiss-split-feedstock/pull/17#issuecomment-739909855
cc @h-vetinari

### CODE REVIEW

**Comment by mbargull** [comment]
(NB: If the same is needed on Windows, then I'll leave it for someone else to address. I.e., I don't have time/nerve for Windows CMD hacking.)

**Comment by jaimergp** [comment]
If needed, I'll handle the Windows part once this PR is merged!

**Comment by h-vetinari** [comment]
This is running into the same gcc-vs-nvcc version problem that I had in https://github.com/conda-forge/faiss-split-feedstock/pull/17 resp. https://github.com/conda-forge/faiss-split-feedstock/pull/19:

* nvcc 9.2 requires gcc <= 7
* nvcc 10.x requires gcc <= 8
* nvcc 11.x requires gcc <= 9 [presumably; not _yet_ relevant]

Possible solutions are [manual downgrade](https://github.com/conda-forge/faiss-split-feedstock/pull/18/commits/ecabd2fc678446c13a545497a19e82a92af78ada) (won't survive rerender), [across-the-bank downgrade](https://github.com/conda-forge/faiss-split-feedstock/pull/17/commits/8e8a3e7c276a3b90bc510706ec2f776d1a6fd483) to gcc=7 on linux, or tackling https://github.com/conda-forge/conda-forge-pinning-feedstock/issues/1000.

File: recipe/install_nvcc.sh
Comment (jakirkham) [inline recipe/install_nvcc.sh:131]: Would it make sense to just loop through `${@}` and see if `-ccbin` is there without the `exec`? Might simplify things a bit 🙂

**Review by jakirkham — COMMENTED** [review]
Thanks for working on this Marcell! 😄

Generally this makes sense. CMake has gotten a bit smarter with setting these things. So it needs less hand-holding, which is nice to see 🙂

Made one suggestion below

File: recipe/install_nvcc.sh
Comment (mbargull) [reply recipe/install_nvcc.sh:131]: > Would it make sense to just loop through `${@}`

This is exactly what `for i` does ;)
> without the `exec`?

The `exec` is just an independent improvement: If we have a launcher script, it makes sense to `exec` the called process so that our shell process can be stopped/replaced.

File: recipe/meta.yaml
Comment (kkraus14) [inline recipe/meta.yaml:32]: Does this have a `run_export`? If so can we do `<= 7` so users can use an older compiler if desired still?

Similarly, it would be good to handle the CUDA 11.0, and 11.1 requirements: https://github.com/conda-forge/nvcc-feedstock/issues/51

Something like:
```suggestion
        - gcc_linux_64 <=7       # [linux64 and (cuda_compiler_version or "").startswith(("9.", "10."))]
        - gcc_linux_64 <=9       # [linux64 and (cuda_compiler_version or "").startswith("11.0")]
        - gcc_linux_64 <=10     # [linux64 and (cuda_compiler_version or "").startswith("11.1")]
```

File: recipe/install_nvcc.sh
Comment (jakirkham) [reply recipe/install_nvcc.sh:131]: Interesting would have expected that to look more like `for i in "${@}"`. It might be clearer to write that (even if `bash` is doing some magic under-the-hood).

Comment (mbargull) [reply recipe/install_nvcc.sh:131]: > (even if `bash` is doing some magic under-the-hood).

It's standard POSIX shell: https://pubs.opengroup.org/onlinepubs/9699919799/utilities/V3_chap02.html#tag_18_09_04_03

> Interesting would have expected that to look more like `for i in "${@}"`. It might be clearer to write that

I don't mind either way -- I can add a commit to change that :).

File: recipe/meta.yaml
Comment (mbargull) [reply recipe/meta.yaml:32]: This is just the compiler version used during the build/test. We only have `run_exports` for the runtimes (`libgcc-ng`/`libstdcxx-ng`) but not for the compilers themselves. You should be able to build downstream packages with older compilers either way.
I've used the exact versions because that's how we build packages on conda-forge generally, with a fixed compiler version (`7` or `9` recently). I've also not special cased `nvcc=10.1|10.2` to have a consistent version with other packages (we hadn't pinned GCC `8` before).

My goal was to do the least necessary for the build and rather wait until someone fixes the mentioned conda-build bug instead of investing time in workarounds.
That said, I think it's best if I split this change off into another PR and just use a temporary workaround in the CI files for the intended change of this PR.

Comment (mbargull) [reply recipe/meta.yaml]: I hadn't looked at the whole recipe until now...
I don't really understand the purpose of the `{{ compiler('c') }}` dependency in `host` here. It is not used in any way.
Also the other duplicate deps for sysroot/glibc I dont understand either.

I can't deduce what the intentions of the dependency declarations overall are. They need some cleanup, but without knowing the indented use, I can't help with that.

**Comment by mbargull** [comment]
> If needed, I'll handle the Windows part once this PR is merged!

Thanks @jaimergp, much appreciated!

**Comment by mbargull** [comment]
@conda-forge/nvcc, this is ready for review.

I opted to postpone the unrelated GCC version pinning to a later PR (<-- not opened yet, because I had problems understanding what the desired actions would be, see https://github.com/conda-forge/nvcc-feedstock/pull/54#discussion_r537824637 ).

File: recipe/install_nvcc.sh
Comment (jakirkham) [reply recipe/install_nvcc.sh]: It would make it clearer I think. We already are `cat`ing into a script (my fault 😞). So it's already a bit confusing. The less confusing we can make it the better 🙂

File: recipe/install_nvcc.sh
Comment (kkraus14) [inline recipe/install_nvcc.sh:133]: If we find `-ccbin` in the arguments, will this execute `nvcc` twice? The first time within thise loop, and then again after the loop?

File: recipe/meta.yaml
Comment (kkraus14) [reply recipe/meta.yaml]: `nvcc` uses the host compiler during it's compilation. Depending on the CUDA version, there's compatibilities with the version of the host compiler used that we need to respect.

Additionally, as of CUDA 11, CentOS 6 is no longer supported so we needed a way to force only CentOS 7+.

File: recipe/install_nvcc.sh
Comment (mbargull) [reply recipe/install_nvcc.sh:133]: `exec` replaces the current process (i.e., the wrapper script) with the actual `nvcc`.

File: recipe/meta.yaml
Comment (mbargull) [reply recipe/meta.yaml]: > `nvcc` uses the host compiler during it's compilation. Depending on the CUDA version, there's compatibilities with the version of the host compiler used that we need to respect.

This recipe only creates wrapper/(de-)activation scripts => the compiler is not used. If it's meant to be declared for downstream packages, something has to be declared in `requirements/run`.

> Additionally, as of CUDA 11, CentOS 6 is no longer supported so we needed a way to force only CentOS 7+.

The duplication of that in `host`/`run`/`test/requires` is what I meant.

File: recipe/install_nvcc.sh
Comment (jakirkham) [reply recipe/install_nvcc.sh:133]: Yeah it shouldn't. Though agree it might be clearer if we had a simple `if...else` here.

Comment (mbargull) [reply recipe/install_nvcc.sh:133]: To me all of this looks very clear -- but well, I've just messed about with POSIX-y shell code a lot the last few years with all these packaging stuff...
If there are code comments or changes you'd like to have which would help comprehension for the maintainers, i.e., you, please add
````
```suggestion
```
````
blocks I can apply.
(If you're unsure about shell syntax, then I can also just translate pseudo code on the fly if you give me some.)

Comment (kkraus14) [reply recipe/install_nvcc.sh:133]: TIL about `exec`. This is fine @mbargull, resolving this.

**Review by kkraus14 — APPROVED** [review]

**Comment by kkraus14** [comment]
Will let @jakirkham take another look as well before merging

**Comment by kkraus14** [comment]
Merging as no response from others.

**Pull request merged**
